server/api/v1/instance_service.go

```diff
@@ -248,18 +248,156 @@ func (s *InstanceService) UndeleteInstance(ctx context.Context, request *v1pb.Un
 }
 
 // AddDataSource adds a data source to an instance.
-func (*InstanceService) AddDataSource(_ context.Context, _ *v1pb.AddDataSourceRequest) (*v1pb.Instance, error) {
-	return nil, status.Errorf(codes.Unimplemented, "method AddDataSource not implemented")
+func (s *InstanceService) AddDataSource(ctx context.Context, request *v1pb.AddDataSourceRequest) (*v1pb.Instance, error) {
+	if request.DataSources == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "data sources is required")
+	}
+	// We only support add RO type datasouce to instance now, see more details in instance_service.proto.
+	if request.DataSources.Type != v1pb.DataSourceType_READ_ONLY {
+		return nil, status.Errorf(codes.InvalidArgument, "only support add read-only data source")
+	}
+
+	dataSourceMsg, err := convertToDataSourceMessage(request.DataSources)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to convert data source")
+	}
+
+	environmentID, instanceID, err := getEnvironmentInstanceID(request.Instance)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, err.Error())
+	}
+	instance, err := s.store.GetInstanceV2(ctx, &store.FindInstanceMessage{
+		EnvironmentID: &environmentID,
+		ResourceID:    &instanceID,
+		ShowDeleted:   true,
```

Comment:
```suggestion
```

Get it ShowDeleted.

Reply:
// We will always return the resource regardless of its deleted state.
	find.ShowDeleted = true